tools/graph_loading/construct_graph.py

```diff
@@ -0,0 +1,680 @@
+        for i, (src_ids, dst_ids, part_data) in return_dict.items():
+            type_src_ids[i] = src_ids
+            type_dst_ids[i] = dst_ids
+            for feat_name in part_data:
+                if feat_name not in type_edge_data:
+                    type_edge_data[feat_name] = [None] * len(return_dict)
+                type_edge_data[feat_name][i] = part_data[feat_name]
+
+        type_src_ids = np.concatenate(type_src_ids)
+        type_dst_ids = np.concatenate(type_dst_ids)
+        assert len(type_src_ids) == len(type_dst_ids)
+
+        for feat_name in type_edge_data:
+            type_edge_data[feat_name] = np.concatenate(type_edge_data[feat_name])
+            assert len(type_edge_data[feat_name]) == len(type_src_ids)
+
+        edge_type = tuple(edge_type)
+        edges[edge_type] = (type_src_ids, type_dst_ids)
+        # Some edge types don't have edge data.
+        if len(type_edge_data) > 0:
+            edge_data[edge_type] = type_edge_data
+
+    return edges, edge_data
+
+if __name__ == '__main__':
+    argparser = argparse.ArgumentParser("Preprocess graphs")
+    argparser.add_argument("--conf_file", type=str, required=True,
+            help="The configuration file.")
+    argparser.add_argument("--num_processes", type=int, default=1,
+            help="The number of processes to process the data simulteneously.")
+    argparser.add_argument("--output_dir", type=str, required=True,
+            help="The path of the output data folder.")
+    argparser.add_argument("--graph_name", type=str, required=True,
+            help="The graph name")
+    argparser.add_argument("--remap_node_id", type=bool, default=False,
+            help="Whether or not to remap node IDs.")
+    argparser.add_argument("--output_format", type=str, default="DistDGL",
+            help="The output format of the constructed graph.")
+    args = argparser.parse_args()
+    num_processes = args.num_processes
+    process_confs = json.load(open(args.conf_file, 'r'))
+
+    node_id_map, node_data = process_node_data(process_confs['node'], args.remap_node_id)
+    edges, edge_data = process_edge_data(process_confs['edge'], node_id_map)
+    num_nodes = {}
+    for ntype in set(list(node_data.keys()) + list(node_id_map.keys())):
+        # If a node type has Id map.
+        if ntype in node_id_map:
+            num_nodes[ntype] = len(node_id_map[ntype])
+        # If a node type has node data.
+        elif ntype in node_data:
+            for feat_name in node_data[ntype]:
+                # Here we only need to look at the length of the first node features
+                # to get the number of nodes for the node type.
+                num_nodes[ntype] = len(node_data[ntype][feat_name])
+                break
+        else:
+            # A node type must have either ID map or node data.
+            raise ValueError('Node type {} must have either ID map or node data'.format(ntype))
+    g = dgl.heterograph(edges, num_nodes_dict=num_nodes)
```

Comment:
Are we going to add reverse edges?

Reply:
here i don't. we can add an argument to add reverse edges.